Userland/useradd.cpp

```diff
@@ -0,0 +1,111 @@
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/CArgsParser.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+static constexpr uid_t BASE_UID = 1000;
+static constexpr gid_t USERS_GID = 100;
+const char* DEFAULT_SHELL = "/bin/sh";
+
+int main(int argc, char** argv)
+{
+    CArgsParser args_parser("useradd");
+    args_parser.add_arg("d", "home_dir", "home directory for the new user");
+    args_parser.add_arg("u", "uid", "user ID (uid) for the new user");
+    args_parser.add_arg("g", "group", "group ID (gid) for the new user");
+    args_parser.add_arg("m", "create home directory if it does not exist");
+    args_parser.add_arg("s", "shell", "name of default shell binary for the new user");
+    args_parser.add_arg("n", "name", "gecos name of the new user");
+    CArgsParserResult args = args_parser.parse(argc, argv);
+
+    if (argc <= 1) {
+        args_parser.print_usage();
+        return 0;
+    }
+
+    const char* username = argv[argc - 1]; // Always the user name!
+    uid_t uid = BASE_UID;
+    gid_t gid = USERS_GID;
+
+    // First, let's sort out the uid for the user
+    if (args.is_present("u")) {
+        bool ok;
+        uid = static_cast<uid_t>(args.get("u").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid uid %s\n", args.get("u").characters());
+            return 3;
+        }
+
+        if (getpwuid(uid)) {
+            fprintf(stderr, "uid %u already exists!\n", uid);
+            return 4;
+        }
+
+    } else {
+        uid = BASE_UID;
+        for (; getpwuid(uid); uid++) {
+        }
+    }
+
+    // Now let's sort out the gid!
+    if (args.is_present("g")) {
+        bool ok;
+        gid = static_cast<gid_t>(args.get("g").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid gid %s\n", args.get("g").characters());
+            return -1;
+        }
+    }
+
+    String shell = args.is_present("s") ? args.get("s") : String(DEFAULT_SHELL);
+    String home = args.is_present("d") ? args.get("d") : String("/");
+    String gecos = args.is_present("n") ? args.get("n") : String("Anonymous");
+
+    if (args.is_present("m")) {
+        StringBuilder builder;
+        builder.append("/home/");
+        builder.append(username);
+
+        String home_path(builder.to_string());
+        if (mkdir(home_path.characters(), 0700)) {
+            perror("mkdir");
+            return 12;
+        }
+
+        if (chown(home_path.characters(), uid, gid)) {
+            perror("chown");
+            return 12;
```

Comment:
So say chown here fails, so we bail out and leave the directory in place. The administrator, confused, runs `useradd` again, and this time is cannot create the home directory because it already exists.

Maybe we should either try to remove the directory on error, or tolerate the directory already existing above?

Reply:
Maybe it would also make sense to check if we're root (or if we're able to open `/etc/passwd`) first, to minimize chances of ever hitting these errors (and leaving the directory around).

Reply:
Good point. I'll move the `fopen` for `pwfile` above this.

Reply:
To add to this, if they run `useradd` again, then they'll get `file already exists` (paraphrasing) so it'll be up to them to manually remove it. imho, considering they're already in the terminal making them have to run `rm -rf` doesn't seem like too much of an inconvenience.

Reply:
The inconvenience is not doing an `rmdir`, it's that:
* They didn't create the directory explicitly (`useradd` did it for them)
* `useradd` appeared to _not work_, but has had an effect. You'd usually want something to "either work on not". Think syscalls: if you get an `-ESOMETHING`, you expect it not to change *anything* (except maybe some syscall counter). It'd be wrong for a syscall to do half of the work and then return an error half-way without cleaning up the first half. Of course, userspace utilities are not syscalls, but we can at least try to behave like that.
* If the error messages don't specify *what* is it that it failed to `chown()` and/or `mkdir()`, the admin will have no idea what directory shey should even look at, and will have to resort to `strace`'ing `useradd` to see what it is that it doesn't like. Trust me, I've been there :smile: So we should try to write code that doesn't make users resort to that.